common/testing/taskpoller/taskpoller.go

```diff
@@ -0,0 +1,313 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package taskpoller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	failurepb "go.temporal.io/api/failure/v1"
+	historypb "go.temporal.io/api/history/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/debug"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/rpc"
+	"go.temporal.io/server/common/testing/testvars"
+	"go.temporal.io/server/service/history/consts"
+)
+
+type (
+	TaskPoller struct {
+		t         *testing.T
+		logger    log.Logger
+		client    workflowservice.WorkflowServiceClient
+		namespace string
+	}
+	Options struct {
+		ctx                 context.Context
+		tv                  *testvars.TestVars
+		maxRetries          int
+		pollStickyTaskQueue bool
+	}
+	// OptionFunc is a function to change an Options instance
+	OptionFunc func(*Options)
+)
+
+var (
+	// DrainWorkflowTask returns an empty RespondWorkflowTaskCompletedRequest response
+	DrainWorkflowTask = func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondWorkflowTaskCompletedRequest, error) {
+		return &workflowservice.RespondWorkflowTaskCompletedRequest{}, nil
+	}
+	// WithTimeout defines a timeout for a task poller method (includes *all* RPC calls it has to make)
+	WithTimeout = func(timeout time.Duration) OptionFunc {
+		return func(o *Options) {
+			o.ctx, _ = rpc.NewContextWithTimeoutAndVersionHeaders(timeout * debug.TimeoutMultiplier)
```

Comment:
I understand what you are trying to achieve here, but I think you need `o.timeout` instead of `o.ctx`.

Reply:
> You can't drop cancel. It is a memory leak.

Yup, I thought that's alright in this context (ie testing). We've been using `testcore.NewContext()` for years, and it drops the cancel.

> I think you need o.timeout instead of o.ctx.

Well, this is a single timeout for the entire function (ie including _all_ RPCs). If I create a new context from the timeout for each RPC, that's a different (and not desirable, I think).